Test multiple chunks in formulate reorg

jflatow opened the pull request.

jflatow committed:
Test multiple chunks in formulate reorg

jflatow committed:
Fix broken StorageLock...

The StorageLocks all have a default deadline, they don't just remain locked.
For some reason (yet to debug), the worker calls are taking a long time / timing out, and due to the broken lock additional worker processes are piling up.

jflatow committed:
Dont give up formulating a reorg, even if we are missing local data

hayesgm commented on pallets/cash/src/internal/events.rs:
```diff
@@ -84,18 +86,18 @@ pub fn risk_adjusted_value<T: Config>(
-        _ => Err(Reason::WorkerBusy),
-    };
-    result
+    if TRACK_CHAIN_EVENTS_LOCK.fetch_add(1, Ordering::SeqCst) == 0 {
```
Curious if this provides _different_ guarantees than the previous lock, regardless of the timeout issue? Does the `StorageLock` use the same underlying mechanism?

jflatow replied:
Yeah I guess this only worked for native since the memory isn't shared by WASM threads, reverted to 120s deadline with storage lock

hayesgm reviewed (approved):
looks good; just wondering if there was another change in regards to unavailable data for re-organization.

jflatow committed:
Back to StorageLock with increased deadline

Each WASM worker runs in its own thread and does not have shared memory, so the previous fix only worked for native.

Setting to 120s deadline should work most of the time, and when it doesn't it's not the end of the world to ignore a few irrelevant blocks, just inefficient if it happens too often.
The other side effect is that the storage lock persists across restarts, which is the downside of making it too long.

jflatow committed:
Rewrite formulate reorg to process forward/reverse in lockstep

Also fix the handling for traversing before the first block, but err out if the worker is otherwise missing data.

coburncoburn reviewed (approved):
🧼 this looks much tidier and therefore much more likely to work correctly 👏

hayesgm reviewed (approved):
lgtm

hayesgm committed:
Bump version
